### PR TITLE
[SPARK-3533][Core] Add saveAsTextFileByKey() method to RDDs

### PR DESCRIPTION
**Pull request by ilganeli**
This patch adds a method to allow saving an RDD as multiple text files split up by key. I've included a test suite that will verify its functionality. For now the test does not check success, it merely executes the functions. This initial submission is intended to test whether a run-time error was solved.

### CODE REVIEW

**Comment by ilganeli** [comment]
retest this please

**Comment by nchammas** [comment]
Thanks for opening this PR! Are you planning to add matching methods/tests for the Python and Java APIs?

**Comment by ilganeli** [comment]
Hi @nchammas - I'd be happy to add tests for the others as soon as I can figure out this init() error. I'm open to suggestions. Thanks!

**Comment by ilganeli** [comment]
retest this please

**Comment by ilganeli** [comment]
Hi @nchammas, I believe I've resolved the issues I had with the code and I've added a test case to the RDDPairFunctionsSuite that checks that the files were created and validates their contents.

**Comment by ilganeli** [comment]
retest this please

**Comment by ilganeli** [comment]
All - any idea why this would be failing? I don't see any specific compilation errors. Attempting to build locally, I see some strange failures both in this branch and in the trunk that showed up after a master merge. Is there any reason that the changes I've added would have broken anything?

**Comment by srowen** [comment]
```
[error] /home/jenkins/workspace/SparkPullRequestBuilder/core/src/test/scala/org/apache/spark/api/python/PythonRDDSuite.scala:71: class RDDMultipleTextOutputFormat takes type parameters
[error]       classOf[RDDMultipleTextOutputFormat].toString,
[error]  
```

**Comment by ilganeli** [comment]
Thanks, Sean - stupid question, how did you find that? I don't see anything useful in the console output or  build info.

**Comment by ilganeli** [comment]
retest this please

**Comment by srowen** [comment]
@ilganeli I just look at the full console output and search for [error]

**Comment by ilganeli** [comment]
All - I attempted to add an API for Python and a test for it as well but I ran into issues I couldn't figure out with class conversion. The way things are implemented for saveAsHadoopFile, there's some confusing reflection going on and when I attempted to replicate it using the custom RDDMultipleTextOutputFormat class as the OutputFormat class, I experienced stage failures when running the tests. I don't think I'm capable of implementing the Python side of things, the commit with most of the Python code necessary is here: https://github.com/apache/spark/commit/c04efeae2728f663a3848be4a526de480e121185.

**Comment by ilganeli** [comment]
retest this please

**Comment by ilganeli** [comment]
@srowen @nchammas What all does this need to get it to a ready state? Thanks.

**Comment by nchammas** [comment]
Hey @ilganeli, I'm not sure since this is outside my comfort zone. Pinging @pwendell who may be able to direct you better.

**Comment by srowen** [comment]
My entirely personal opinion is I'm neutral on whether this is worth more API methods.

**Comment by nchammas** [comment]
@srowen [SPARK-3533](https://issues.apache.org/jira/browse/SPARK-3533) has a lot of votes and watchers, and there are a few linked questions on Stack Overflow from there, the most popular one being [this question](http://stackoverflow.com/q/23995040/877069), which has 12 upvotes ATM and close to 4,000 views in about a year, as well as several linked questions asking about the same thing.

From a user perspective, I can definitely say that this is a sought-after method.

**Comment by ilganeli** [comment]
@srowen  @pwendell Nicholas correctly points out that there's a substantial amount of interest for this. Independent of the community interest, I think this is a useful functionality since it allows a quick and dirty organization of data on disk by key. There's not really a simple way to do this with existing APIs for someone writing this from scratch.

**Comment by ilganeli** [comment]
Anyone? I just hear crickets :)

**Comment by ilganeli** [comment]
Folks, anything on this?

**Comment by ilganeli** [comment]
cc @pwendell @JoshRosen @andrewor14 Does anyone feel that this work is valuable? I'd hate to close the PR when there's so much community interest.

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala [hunk 1]
```diff
@@ -21,6 +21,10 @@ import java.nio.ByteBuffer
 import java.text.SimpleDateFormat
 import java.util.{Date, HashMap => JHashMap}
 
+import org.apache.commons.lang.ClassUtils
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Move these after "import scala.xxx"

**Comment by davies** [comment]
+1 for this feature, it's very useful, DPark also has this in early stage [https://github.com/douban/dpark/commit/6e62b76ef8d80d8a8f4c55378a6471f72c0ec339]. 

Should we also have a simple API for text file? and some options to do compress?

The title has `PySpark`, but there is no Python API right now, could you also add it?

**Comment by ilganeli** [comment]
@davies I spent a good amount of time trying to get Python bindings to work. In the Python API for ```saveAsHadoopFile``` there's a lot of confusing reflection going on, however, and I wasn't able to figure it out.

**Comment by davies** [comment]
@ilganeli Could you remove `PySpark` from the title and create a JIRA for the missing Python API?

**Comment by ilganeli** [comment]
@davies I've created a JIRA for the Python API here along with some starter code.
https://issues.apache.org/jira/browse/SPARK-6780

**Comment by ilganeli** [comment]
All - anything to do here?

**Comment by nchammas** [comment]
@ilganeli - I think this is a useful feature, but given that there isn't a strong committer sponsor (I am just a random contributor) to take this PR through to the end, I suggest closing it for now and perhaps converting it into a package for [Spark Packages](http://spark-packages.org/).

If this feature proves to work well and is popular as an external package, it will probably become much easier then to merge it into core in an expedient manner.

What do you think?

**Comment by ilganeli** [comment]
Closing this PR, I'll move it to Packages at some point.

**Comment by brendancol** [comment]
hey did this ever make it into Packages?  Great idea and would love to use it.

**Comment by ilganeli** [comment]
Did not - I can port it over if you need it. Thanks !



Thank you,
Ilya Ganelin



-----Original Message-----

**Comment by brendancol** [comment]
Hey great. yeah just looking for this functionality on the python side.  I'm currently doing n+1 select statements to do a multiple outputs from a dataframe, looked like this PR was the furtherest folks had come in having a saveAsTextFileByKey().  I have a stackoverflow (http://stackoverflow.com/questions/30673144/python-spark-dataframes-better-way-to-export-groups-to-text-file) question about this issue.  If there is a simpler way of achieving this, I would love to know as I'm a bit of a noob.  Thanks

**Comment by baconmania** [comment]
Couldn't find this on Packages... any chance this'll get ported?

**Comment by silasdavis** [comment]
I am trying to get this re-opened, please see: https://issues.apache.org/jira/browse/SPARK-3533

I have a prototype implementation that allows you to write multiple outputs using Hadoop 2 multiple outputs for arbitrary OutputFormats not just Text (I have implemented saveAsMultipleTextFiles, saveAsMultipleAvroFiles, and saveAsMultipleParquetFiles, but other underlying ouput formats should be straight-forward) linked to from the ticket at this gist: https://gist.github.com/silasdavis/d1d1f1f7ab78249af462. It works with the current Spark API and doesn't require any changes but I think it would be better to integrate it. It looks like we need to get a Spark committer interested to drive it forward.

**Pull request closed without merging**